File: LeetCodePractice/src/BFS/PacificAtlanticWaterFlow.java
package BFS;

import java.util.*;

/*Given an m x n matrix of non-negative integers representing the height of each unit cell in a continent, the "Pacific ocean" touches the left and top edges of the matrix and the "Atlantic ocean" touches the right and bottom edges.

Water can only flow in four directions (up, down, left, or right) from a cell to another one with height equal or lower.

Find the list of grid coordinates where water can flow to both the Pacific and Atlantic ocean.

Note:
The order of returned grid coordinates does not matter.
Both m and n are less than 150.
Example:

Given the following 5x5 matrix:

  Pacific ~   ~   ~   ~   ~ 
       ~  1   2   2   3  (5) *
       ~  3   2   3  (4) (4) *
       ~  2   4  (5)  3   1  *
       ~ (6) (7)  1   4   5  *
       ~ (5)  1   1   2   4  *
          *   *   *   *   * Atlantic

Return:

[[0, 4], [1, 3], [1, 4], [2, 2], [3, 0], [3, 1], [4, 0]] (positions with parentheses in above matrix).*/
public class PacificAtlanticWaterFlow {
	int[] dirs = { 0, 1, 0, -1, 0 };
	int rows, cols;

	public List<int[]> pacificAtlantic(int[][] matrix) {
		List<int[]> res = new ArrayList<>();
		if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
			return res;
		}
		rows = matrix.length;
		cols = matrix[0].length;
		Queue<int[]> pacificQueue = new LinkedList<>();
		Queue<int[]> atlanticQueue = new LinkedList<>();
		boolean[][] pacificVisited = new boolean[rows][cols];
		boolean[][] atlanticVisited = new boolean[rows][cols];
		for (int r = 0; r < rows; r++) {
			pacificQueue.offer(new int[] { r, 0 });
			pacificVisited[r][0] = true;
			atlanticQueue.offer(new int[] { r, cols - 1 });
			atlanticVisited[r][cols - 1] = true;
		}
		for (int c = 0; c < cols; c++) {
			pacificQueue.offer(new int[] { 0, c });
			pacificVisited[0][c] = true;
			atlanticQueue.offer(new int[] { rows - 1, c });
			atlanticVisited[rows - 1][c] = true;
		}
		bfs(matrix, pacificQueue, pacificVisited);
		bfs(matrix, atlanticQueue, atlanticVisited);		
		for(int r = 0; r < rows; r++) {
			for(int c = 0; c < cols; c++) {
				if(pacificVisited[r][c] && atlanticVisited[r][c]) {
					res.add(new int[] {r, c});
				}
			}
		}
		return res;
	}

	private void bfs(int[][] matrix, Queue<int[]> queue, boolean[][] visited) {
		while (!queue.isEmpty()) {
			int[] cur = queue.poll();
			for (int i = 0; i < dirs.length - 1; i++) {
				int nb_r = cur[0] + dirs[i];
				int nb_c = cur[1] + dirs[i + 1];
				if (nb_r >= 0 && nb_r < rows && nb_c >= 0 && nb_c < cols && !visited[nb_r][nb_c]
						&& matrix[nb_r][nb_c] >= matrix[cur[0]][cur[1]]) {
					queue.offer(new int[] { nb_r, nb_c });
					visited[nb_r][nb_c] = true;
				}
			}
		}
	}
}

File: LeetCodePython/PacificAtlanticWaterFlow.py
"""
There is an m x n rectangular island that borders both the Pacific Ocean and Atlantic Ocean. The Pacific Ocean touches the island's left and top edges, and the Atlantic Ocean touches the island's right and bottom edges.

The island is partitioned into a grid of square cells. You are given an m x n integer matrix heights where heights[r][c] represents the height above sea level of the cell at coordinate (r, c).

The island receives a lot of rain, and the rain water can flow to neighboring cells directly north, south, east, and west if the neighboring cell's height is less than or equal to the current cell's height. Water can flow from any cell adjacent to an ocean into the ocean.

Return a 2D list of grid coordinates result where result[i] = [ri, ci] denotes that rain water can flow from cell (ri, ci) to both the Pacific and Atlantic oceans.



Example 1:


Input: heights = [[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]
Output: [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]
Example 2:

Input: heights = [[2,1],[1,2]]
Output: [[0,0],[0,1],[1,0],[1,1]]


Constraints:

m == heights.length
n == heights[r].length
1 <= m, n <= 200
0 <= heights[r][c] <= 105
"""
import collections


class PacificAtlanticWaterFlow(object):
    def pacificAtlantic(self, heights):
        """
        :type heights: List[List[int]]
        :rtype: List[List[int]]
        """
        rows, cols = len(heights), len(heights[0])
        pacific_queue = collections.deque()
        atlantic_queue = collections.deque()
        for r in range(rows):
            pacific_queue.append((r, 0))
        for c in range(1, cols):
            pacific_queue.append((0, c))
        for r in range(rows):
            atlantic_queue.append((r, cols - 1))
        for c in range(cols - 1):
            atlantic_queue.append((rows - 1, c))

        def bfs(q):
            visited = set(q)
            while q:
                (r, c) = q.popleft()
                for (dr, dc) in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                    if 0 <= r + dr < rows \
                            and 0 <= c + dc < cols \
                            and (r + dr, c + dc) not in visited \
                            and heights[r][c] <= heights[r + dr][c + dc]:
                        q.append((r + dr, c + dc))
                        visited.add((r + dr, c + dc))
            return visited

        return list(bfs(pacific_queue).intersection(bfs(atlantic_queue)))

    def pacificAtlanticDFS(self, heights):
        """
        :type heights: List[List[int]]
        :rtype: List[List[int]]
        """
        if not heights:
            return []
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        rows, cols = len(heights), len(heights[0])
        pacific_visited = [[False for _ in range(cols)] for _ in range(rows)]
        atlantic_visited = [[False for _ in range(cols)] for _ in range(rows)]
        res = []

        def dfs(r, c, visited):
            visited[r][c] = True
            for dir in dirs:
                nb_r, nb_c = r + dir[0], c + dir[1]
                if 0 <= nb_r < rows \
                        and 0 <= nb_c < cols \
                        and not visited[nb_r][nb_c] \
                        and heights[r][c] <= heights[nb_r][nb_c]:
                    dfs(nb_r, nb_c, visited)

        for r in range(rows):
            dfs(r, 0, pacific_visited)
            dfs(r, cols - 1, atlantic_visited)
        for c in range(cols):
            dfs(0, c, pacific_visited)
            dfs(rows - 1, c, atlantic_visited)

        for r in range(rows):
            for c in range(cols):
                if pacific_visited[r][c] and atlantic_visited[r][c]:
                    res.append((r, c))
        return res